### PR TITLE
GROM-138: Manual upload UI

### PR DESCRIPTION
**Pull request by dan-0**
Adds ViewModel scaffolding for "Upload" registrations and failed upload UX.

### CODE REVIEW

**Comment by dan-0** [comment]
Screen Shot 2020-05-13 at 8 33 42 PM
Changes

File: .github/workflows/android.yml [hunk 1]
```diff
@@ -20,8 +20,8 @@ jobs:
       - name: Uninstall NDK
         run: sudo ${ANDROID_HOME}/tools/bin/sdkmanager --uninstall `ndk-bundle`
 
-      - name: Build with Gradle
-        run: ./gradlew build
```
Comment (dan-0) [inline .github/workflows/android.yml:24]: This failed, it was unable to build a production variant due to a dependency issue. I changed this out for the internalDebug variant until we get closer to completion.

Comment (huttneab) [reply .github/workflows/android.yml:24]: 👍

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java [hunk 2]
```diff
@@ -74,6 +86,50 @@ protected void onCreate(Bundle savedInstanceState) {
...
+
+    private void observeState() {
+        viewModel.getState().observe(this, mainActivityState -> {
+
+            // Reset base states
+            upploadButton.setEnabled(true);
+            failedRegistrationsContainer.setVisibility(View.GONE);
+            failedRegistrations.setText("");
+
+            if (mainActivityState instanceof MainActivityState.Content) {
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java:106]: this looks prettier in kotlin

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java [hunk 3]
```diff
@@ -74,6 +86,50 @@ protected void onCreate(Bundle savedInstanceState) {
...
+                MainActivityState.Content content = (MainActivityState.Content) mainActivityState;
+
+                String pendingUploads = String.format(
+                        Locale.getDefault(),
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java:110]: Not sure if this is just needed in order to format date, but we don't have to worry about language on this page, only within the registration process. This is always english.

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java [hunk 4]
```diff
@@ -74,6 +86,50 @@ protected void onCreate(Bundle savedInstanceState) {
...
+                        "%d",
+                        content.getPendingUploads()
+                );
+
+                pendingRegistrations.setText(pendingUploads);
+
+                if (content.getFailedUploads() > 0) {
+                    failedRegistrationsContainer.setVisibility(View.VISIBLE);
+
+                    String failedUploads = String.format(
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java:120]: same same but different

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                pendingUploads = requests.size,
+                failedUploads = 0 // TODO need to figure out how we're going to calculate this
+            )
+
+            updateState(content)
+        }
+    }
+
+    fun uploadRegistrations() {
+        updateState(MainActivityState.Loading)
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt:47]: god this is so much cleaner than the old upload service class, mainly due to storing it in a pre-formatted json blob, but still.

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                    }
+                }
+            }
+
+            resultMap.forEach {
+                // TODO Handle observable results
+            }
+
+            // TODO populate this with real values
+            val result = MainActivityState.Content(
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt:70]: lol, hard code for success!

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        _state.postValue(newState)
+    }
+
+    override fun onCleared() {
+        super.onCleared()
+        supervisorJob.cancelChildren()
+    }
+}
+
+class MainActivityViewModelFactory(
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt:95]: what's the reason behind the factory?

File: app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package com.rockthevote.grommet.util
+
+import kotlinx.coroutines.CoroutineDispatcher
+
+/**
+ * Provides Coroutine dispatchers for better control within the application and tests
+ */
+interface DispatcherProvider {
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt:8]: didn't we already have  a DispatcherProvider interface?

Comment (huttneab) [reply app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt:8]: ah, you moved it from the coroutines folder, ya?

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java (hunk 3 above)
Comment (dan-0) [reply app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java:110]: Eh, it was a lint warning so 🤷‍♂️, not needed, but not harmful

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt (hunk 7 above)
Comment (dan-0) [reply app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt:95]: It's necessary to provide dependencies to the ViewModel when instantiated in the MainActivity via ViewModelProvider(...).get(...)

File: app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt (hunk 8 above)
Comment (dan-0) [reply app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt:8]: I swore we did, but I can't find it.

File: app/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -43,6 +43,10 @@
     <string name="validating_partner_id" translatable="false">Validating Partner ID</string>
     <string name="shift_progress" translatable="false">Shift Progress</string>
     <string name="pending_registrations">Applications Pending Upload</string>
+
+    <!--Wording here fits the requirement that we don't make a user "worry" about a failed upload-->
+    <string name="failed_registration_uploads">Requires Admin For Upload</string>
```
Comment (huttneab) [inline app/src/main/res/values/strings.xml:48]: needs `translatable="false"`

File: app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt (hunk 8 above)
Comment (dan-0) [reply app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt:8]: You know what, it was in the other PR I think. I see it now, I'm removing this.

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java (hunk 3 above)
Comment (huttneab) [reply app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java:110]: works for me

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt (hunk 7 above)
Comment (huttneab) [reply app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt:95]: 👍

File: app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt (hunk 8 above)
Comment (huttneab) [reply app/src/main/java/com/rockthevote/grommet/util/DispatcherProvider.kt:8]: 👍

File: app/src/main/res/values/strings.xml (hunk 9 above)
Comment (dan-0) [reply app/src/main/res/values/strings.xml:48]: 👍 Should `translatable="false"` be in the strings on either side of this as well?

Comment (huttneab) [reply app/src/main/res/values/strings.xml]: ya, but technically out of scope

**Pull request merged**